ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,147 @@
+package io.confluent.ksql.internal;
+
+import io.confluent.ksql.engine.QueryEventListener;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.QueryMetadata;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.processor.ThreadMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class UtilizationMetricsListener implements Runnable, QueryEventListener {
+
+    private final String STREAM_THREAD_GROUP = "stream-thread-metrics";
+    private final String THREAD_ID = "thread-id";
+
+    private final List<KafkaStreams> kafkaStreams;
+    private final Logger logger = LoggerFactory.getLogger(UtilizationMetricsListener.class);
+    private final List<String> metrics;
+    private final Time time;
+    private final double windowSize;
+
+    private final Map<String, Double> previousPollTime;
+    private final Map<String, Double> previousRestoreConsumerPollTime;
+    private final Map<String, Double> previousSendTime;
+    private final Map<String, Double> previousFlushTime;
+
+    public UtilizationMetricsListener(final double windowSize){
+        this.kafkaStreams = new ArrayList<>();
+        this.metrics = new LinkedList<>();
+        // we can add these here or pass it in through the constructor
+        metrics.add("poll-time-total");
+        metrics.add("restore-consumer-poll-time-total");
+        metrics.add("send-time-total");
+        metrics.add("flush-time-total");
+        time = Time.SYSTEM;
+        this.windowSize = windowSize;
+        previousPollTime = new HashMap<>();
+        previousRestoreConsumerPollTime = new HashMap<>();
+        previousSendTime = new HashMap<>();
+        previousFlushTime = new HashMap<>();
+    }
+
+    @Override
+    public void onCreate(
+            final ServiceContext serviceContext,
+            final MetaStore metaStore,
+            final QueryMetadata queryMetadata) {
+        kafkaStreams.add(queryMetadata.getKafkaStreams());
+    }
+
+    @Override
+    public void onDeregister(final QueryMetadata query) {
+        // Question - if we terminate a query and then restart it, will the underling
```

Comment:
we would not reuse the name on a terminate (it gets a new ID), so we should be good here. We do reuse the name on an upgrade. I think we should always just clear up all our samples for the query here. It's not a big deal if we miss a window because a user upgraded their query.